File: dis-ingest/src/test/java/io/pivotal/dis/ingest/service/store/AmazonS3FileStoreTest.java
package io.pivotal.dis.ingest.service.store;

import com.amazonaws.services.s3.AmazonS3;
import com.amazonaws.services.s3.model.AccessControlList;
import com.amazonaws.services.s3.model.GroupGrantee;
import com.amazonaws.services.s3.model.Permission;
import com.amazonaws.services.s3.model.PutObjectRequest;
import org.apache.commons.io.IOUtils;
import org.junit.Test;

import java.io.IOException;
import java.io.InputStream;
import java.lang.reflect.InvocationHandler;
import java.lang.reflect.Proxy;
import java.util.HashMap;
import java.util.Map;
import java.util.function.Consumer;

import static org.hamcrest.Matchers.*;
import static org.junit.Assert.assertThat;

public class AmazonS3FileStoreTest {

    @Test
    public void savesTimestampedFileToS3WithPubliclyReadableAcl() {
        Map<String, Object> params = new HashMap<>();
        AmazonS3 mockAmazonS3 = proxy(AmazonS3.class, "putObject", args -> {
            PutObjectRequest putObjectRequest = (PutObjectRequest) args[0];
            params.put("bucketName", putObjectRequest.getBucketName());
            params.put("key", putObjectRequest.getKey());
            params.put("input", toString(putObjectRequest.getInputStream()));
            params.put("acl", putObjectRequest.getAccessControlList());
        });
        AccessControlList publicReadableAcl = new AccessControlList();
        publicReadableAcl.grantPermission(GroupGrantee.AllUsers, Permission.Read);
        AmazonS3FileStore fileStore = new AmazonS3FileStore(mockAmazonS3, "bucketName", publicReadableAcl);
        fileStore.save("supplied filename", "some stuff");

        assertThat(params, allOf(
                hasEntry("input", "some stuff"),
                hasEntry("bucketName", "bucketName"),
                hasEntry("key", "supplied filename"),
                hasKey("acl")
        ));

        AccessControlList acl = (AccessControlList) params.get("acl");
        assertThat(acl, equalTo(publicReadableAcl));
    }

    @Test
    public void savesTimestampedFileToS3WithDefaultAcl() {
        Map<String, Object> params = new HashMap<>();
        AmazonS3 mockAmazonS3 = proxy(AmazonS3.class, "putObject", args -> {
            PutObjectRequest putObjectRequest = (PutObjectRequest) args[0];
            params.put("bucketName", putObjectRequest.getBucketName());
            params.put("key", putObjectRequest.getKey());
            params.put("input", toString(putObjectRequest.getInputStream()));
            params.put("acl", putObjectRequest.getAccessControlList());
        });
        AccessControlList defaultAcl = new AccessControlList();
        AmazonS3FileStore fileStore = new AmazonS3FileStore(mockAmazonS3, "bucketName", defaultAcl);
        fileStore.save("supplied filename", "some stuff");

        assertThat(params, allOf(
                hasEntry("input", "some stuff"),
                hasEntry("bucketName", "bucketName"),
                hasEntry("key", "supplied filename"),
                hasKey("acl")
        ));


        AccessControlList acl = (AccessControlList) params.get("acl");
        assertThat(acl, equalTo(defaultAcl));
    }

    private String toString(InputStream stream) {
        try {
            return IOUtils.toString(stream);
        } catch (IOException e) {
            return e.toString();
        }
    }

    private <T> T proxy(Class<T> interfaceClass, String methodName, Consumer<Object[]> handler) {
        InvocationHandler invocationHandler = (proxy, method, args) -> {
            if (method.getName().equals(methodName)) {
                handler.accept(args);
            }
            return null;
        };
        return interfaceClass.cast(Proxy.newProxyInstance(interfaceClass.getClassLoader(), new Class[]{interfaceClass}, invocationHandler));
    }

}

File: dis-ingest/src/test/kotlin/io/pivotal/dis/ingest/test/AmazonS3FileStoreTest.kt
package io.pivotal.dis.ingest.test

import com.amazonaws.services.s3.AmazonS3
import com.amazonaws.services.s3.model.AccessControlList
import com.amazonaws.services.s3.model.S3Object
import com.amazonaws.services.s3.model.S3ObjectInputStream
import io.pivotal.dis.ingest.app.store.AmazonS3FileStore
import org.apache.commons.io.IOUtils
import org.junit.Test
import org.mockito.Mockito.*
import org.junit.Assert.*;
import org.hamcrest.Matchers.*;

class AmazonS3FileStoreTest {

    @Test
    fun read_returnsStringOfRequestedFilesContents() {
        val amazonS3 = mock(AmazonS3::class.java)
        val s3Object = S3Object()
        s3Object.objectContent = S3ObjectInputStream(IOUtils.toInputStream("I'm the content!"), null)

        `when`(amazonS3.getObject("bucketName", "fileName")).thenReturn(s3Object)

        val amazonS3FileStore = AmazonS3FileStore(amazonS3, "bucketName", AccessControlList())
        assertThat(amazonS3FileStore.read("fileName"), equalTo("I'm the content!"))
    }

    @Test
    fun read_whenRequestedFileDoesNotExist_returnsNull() {
        val amazonS3 = mock(AmazonS3::class.java)

        `when`(amazonS3.getObject("bucketName", "fileName")).thenReturn(null)

        val amazonS3FileStore = AmazonS3FileStore(amazonS3, "bucketName", AccessControlList())
        assertThat(amazonS3FileStore.read("fileName"), equalTo(null as String?))
    }
}
